File: packages/typescript/src/mcp.js
/**
 * Client for the unified Tempera MCP gateway (`${issuer}/mcp`): stateless
 * streamable-HTTP JSON-RPC 2.0 aggregating every product MCP server behind
 * namespaced tools (palette_*, tempo_*, cradle_*, remi_*, data_engine_*).
 *
 * Requires a bearer minted for audience `tempera-mcp` with scope `mcp:invoke`
 * (or a central tp_ API key). Mirrored by tempera_sdk.TemperaMcpClient in
 * Python; the Rust crate exposes JSON-RPC body builders instead.
 */

import { TEMPERA_MCP_GATEWAY } from "./surface.js";
import { TemperaMcpError, TemperaSdkError, apiErrorFromResponse } from "./errors.js";

export const MCP_PROTOCOL_VERSION = "2026-07-28";

// SEP-2243 routable headers. A middle box has to be able to route a request
// without parsing its body, so the method carries its subject in `Mcp-Name`:
// the tool for tools/call and prompts/get, the uri for the resources/* methods,
// the task id for the tasks/* ones. A gateway that validates them (tempera-mcp
// does) refuses a request that omits one.
const MCP_NAME_FROM = {
  "tools/call": "name",
  "prompts/get": "name",
  "resources/read": "uri",
  "resources/subscribe": "uri",
  "resources/unsubscribe": "uri",
  "tasks/get": "taskId",
  "tasks/update": "taskId",
  "tasks/cancel": "taskId",
};
const BASE64_HEADER_PREFIX = "=?base64?";
const BASE64_HEADER_SUFFIX = "?=";

/** A header value, Base64-wrapped when it could not survive as one. */
function headerSafe(value) {
  const unsafe =
    value.length > 0 &&
    (/^[ \t]|[ \t]$/.test(value) ||
      // eslint-disable-next-line no-control-regex
      /[^\x20-\x7e]/.test(value) ||
      (value.startsWith(BASE64_HEADER_PREFIX) && value.endsWith(BASE64_HEADER_SUFFIX)));
  if (!unsafe) return value;
  const encoded = Buffer.from(value, "utf8").toString("base64");
  return `${BASE64_HEADER_PREFIX}${encoded}${BASE64_HEADER_SUFFIX}`;
}

function mcpName(method, params) {
  const key = MCP_NAME_FROM[method];
  if (!key) return null;
  const value = params?.[key];
  return typeof value === "string" ? headerSafe(value) : null;
}

export class TemperaMcpClient {
  constructor({ url, auth, bearer, fetch: fetchImpl } = {}) {
    this.url = url ?? auth?.mcpUrl;
    if (!this.url) throw new TemperaSdkError("url is required (e.g. https://api.tempera.dev/mcp)");
    this.auth = auth ?? null;
    this.bearer = bearer ?? null;
    this.fetch = fetchImpl ?? auth?.fetch ?? globalThis.fetch;
    if (!this.fetch) throw new TemperaSdkError("fetch is required");
    this.nextId = 1;
  }

  #resolveBearer() {
    if (this.bearer) return this.bearer;
    if (this.auth) return this.auth.bearerFor("tempera-mcp");
    throw new TemperaSdkError("no MCP credential; pass bearer or a TemperaAuth with an apiKey or tempera-mcp tokens");
  }

  /** Send one JSON-RPC request and return its result (throws TemperaMcpError on rpc errors). */
  async rpc(method, params = undefined) {
    const requestMeta = {
      ...((params ?? {})._meta ?? {}),
        "io.modelcontextprotocol/protocolVersion": MCP_PROTOCOL_VERSION,
        "io.modelcontextprotocol/clientCapabilities": {},
    };
    const requestParams = {
      ...(params ?? {}),
      _meta: requestMeta,
    };
    const response = await this.fetch(this.url, {
      method: "POST",
      headers: {
        // Streamable HTTP requires the client to accept BOTH shapes on a POST:
        // the gateway chooses one JSON body or an SSE stream, and a
        // spec-conformant gateway answers 406 to a client that offers only one.
        accept: "application/json, text/event-stream",
        "content-type": "application/json",
        authorization: `Bearer ${this.#resolveBearer()}`,
        "mcp-protocol-version": MCP_PROTOCOL_VERSION,
        "mcp-method": method,
        ...(mcpName(method, requestParams) === null
          ? {}
          : { "mcp-name": mcpName(method, requestParams) }),
      },
      body: JSON.stringify({ jsonrpc: "2.0", id: this.nextId++, method, params: requestParams }),
    });
    const text = await response.text();
    const parsed = text ? JSON.parse(text) : null;
    if (!response.ok) {
      throw apiErrorFromResponse({
        status: response.status,
        statusText: response.statusText,
        headers: response.headers,
        body: parsed,
        product: "mcpGateway",
        operation: method,
      });
    }
    if (parsed?.error) {
      // Uniform rule (same in Python and Rust): a JSON-RPC error object
      // carries its integer code (0 when absent) and string message; a
      // non-conformant non-object error becomes code 0 with its string form.
      const error = parsed.error;
      const isObject = typeof error === "object";
      throw new TemperaMcpError({
        code: isObject && typeof error.code === "number" ? error.code : 0,
        message: isObject ? (typeof error.message === "string" ? error.message : "MCP error") : String(error),
        data: isObject ? error.data : null,
      });
    }
    return parsed?.result;
  }

  /** Discover the stateless MCP server's capabilities and instructions. */
  initialize({ name = "tempera-sdk", version = "0.13.0" } = {}) {
    return this.rpc("server/discover", {
      _meta: { "io.modelcontextprotocol/clientInfo": { name, version } },
    });
  }

  /**
   * Check gateway liveness over JSON-RPC.
   *
   * @deprecated MCP revision 2026-07-28 removed `ping`; a gateway on that
   * revision answers it with a method-not-found. This helper now sends
   * `server/discover` so existing callers keep working, and will be removed
   * in a future release. Call `initialize()` instead.
   */
  ping() {
    return this.initialize();
  }

  /** List every tool the gateway offers: builtins plus namespaced product tools. */
  async listTools() {
    const result = await this.rpc("tools/list");
    return result?.tools ?? [];
  }

  /** Invoke a tool by name; product tool calls are metered as mcp_invocations. */
  callTool(name, args = {}) {
    return this.rpc("tools/call", { name, arguments: args });
  }

  /** Fetch the caller's identity, workspace, and scopes as seen by the gateway. */
  whoami() {
    return this.callTool("tempera_whoami");
  }

  /** Fetch gateway upstream health for every connected product MCP server. */
  status() {
    return this.callTool("tempera_status");
  }
}

export const MCP_ERROR_CODES = Object.freeze(TEMPERA_MCP_GATEWAY.errorCodes);

File: packages/c/src/mcp.c
/*
 * JSON-RPC 2.0 body builders for the unified Tempera MCP gateway
 * (${issuer}/mcp): stateless streamable-HTTP JSON-RPC over the fixed
 * capability-fabric verb surface.
 *
 * The library is HTTP-less: these builders produce the exact request bodies the
 * gateway expects. POST them at tempera_auth_mcp_url() with an
 * "authorization: Bearer <token>" header (a bearer minted for audience
 * tempera-mcp with scope mcp:invoke, or a central tp_ API key), then feed error
 * responses to tempera_mcp_parse_error().
 */

#include <stdlib.h>
#include <string.h>

#include "internal.h"

void tempera_mcp_builder_init(tempera_mcp_builder *builder)
{
    if (builder != NULL) {
        builder->next_id = 1;
    }
}

static long long take_id(tempera_mcp_builder *builder, long long *out_id)
{
    long long id = builder->next_id;

    builder->next_id++;
    if (out_id != NULL) {
        *out_id = id;
    }
    return id;
}

/* "{"jsonrpc":"2.0","id":<id>,"method":"<method>"" -- no trailing brace. */
static void append_envelope(tempera_buf *buf, long long id, const char *method)
{
    tempera_buf_append(buf, "{\"jsonrpc\":\"2.0\",\"id\":");
    tempera_buf_append_i64(buf, id);
    tempera_buf_append(buf, ",\"method\":\"");
    tempera_buf_append(buf, method);
    tempera_buf_append_char(buf, '"');
}

static char *simple_body(tempera_mcp_builder *builder, const char *method, long long *out_id)
{
    tempera_buf buf;

    if (builder == NULL) {
        return NULL;
    }
    tempera_buf_init(&buf);
    append_envelope(&buf, take_id(builder, out_id), method);
    tempera_buf_append_char(&buf, '}');
    return tempera_buf_finish(&buf);
}

char *tempera_mcp_initialize_body(tempera_mcp_builder *builder,
                                  const char *client_name,
                                  const char *client_version,
                                  long long *out_id)
{
    tempera_buf buf;
    char *name;
    char *version;

    if (builder == NULL || client_name == NULL || client_version == NULL) {
        return NULL;
    }
    name = tempera_json_escape(client_name);
    version = tempera_json_escape(client_version);
    if (name == NULL || version == NULL) {
        tempera_string_free(name);
        tempera_string_free(version);
        return NULL;
    }
    tempera_buf_init(&buf);
    append_envelope(&buf, take_id(builder, out_id), "initialize");
    tempera_buf_append(&buf, ",\"params\":{\"protocolVersion\":\"");
    tempera_buf_append(&buf, TEMPERA_MCP_PROTOCOL_VERSION);
    tempera_buf_append(&buf, "\",\"capabilities\":{},\"clientInfo\":{\"name\":\"");
    tempera_buf_append(&buf, name);
    tempera_buf_append(&buf, "\",\"version\":\"");
    tempera_buf_append(&buf, version);
    tempera_buf_append(&buf, "\"}}}");
    tempera_string_free(name);
    tempera_string_free(version);
    return tempera_buf_finish(&buf);
}

char *tempera_mcp_ping_body(tempera_mcp_builder *builder, long long *out_id)
{
    return simple_body(builder, "ping", out_id);
}

char *tempera_mcp_list_tools_body(tempera_mcp_builder *builder, long long *out_id)
{
    return simple_body(builder, "tools/list", out_id);
}

char *tempera_mcp_call_tool_body(tempera_mcp_builder *builder,
                                 const char *tool_name,
                                 const char *arguments_json,
                                 long long *out_id)
{
    tempera_buf buf;
    char *name;

    if (builder == NULL || tool_name == NULL) {
        return NULL;
    }
    name = tempera_json_escape(tool_name);
    if (name == NULL) {
        return NULL;
    }
    tempera_buf_init(&buf);
    append_envelope(&buf, take_id(builder, out_id), "tools/call");
    tempera_buf_append(&buf, ",\"params\":{\"name\":\"");
    tempera_buf_append(&buf, name);
    tempera_buf_append(&buf, "\",\"arguments\":");
    tempera_buf_append(&buf, arguments_json != NULL ? arguments_json : "{}");
    tempera_buf_append(&buf, "}}");
    tempera_string_free(name);
    return tempera_buf_finish(&buf);
}

char *tempera_mcp_whoami_body(tempera_mcp_builder *builder, long long *out_id)
{
    return tempera_mcp_call_tool_body(builder, "tempera_whoami", NULL, out_id);
}

char *tempera_mcp_status_body(tempera_mcp_builder *builder, long long *out_id)
{
    return tempera_mcp_call_tool_body(builder, "tempera_status", NULL, out_id);
}

void tempera_mcp_error_free(tempera_mcp_error *error)
{
    if (error == NULL) {
        return;
    }
    free(error->message);
    error->message = NULL;
    error->code = 0;
}

bool tempera_mcp_parse_error(const char *body, tempera_mcp_error *out_error)
{
    tempera_json *root;
    const tempera_json *error;
    bool found = false;

    if (out_error == NULL) {
        return false;
    }
    memset(out_error, 0, sizeof(*out_error));
    root = tempera_json_parse(body);
    error = tempera_json_get(root, "error");
    if (error == NULL) {
        tempera_json_free(root);
        return false;
    }
    /*
     * Uniform rule (same in TypeScript, Python, and Rust): a JSON-RPC error
     * object carries its integer code (0 when absent) and string message; a
     * non-conformant non-object error becomes code 0 with its string form.
     */
    switch (error->kind) {
    case TEMPERA_JSON_OBJECT: {
        const char *message = tempera_json_as_str(tempera_json_get(error, "message"));
        long long code = 0;

        (void)tempera_json_as_i64(tempera_json_get(error, "code"), &code);
        out_error->code = code;
        out_error->message = tempera_strdup(message != NULL ? message : "MCP error");
        found = true;
        break;
    }
    case TEMPERA_JSON_STRING:
        out_error->message = tempera_strdup(error->text);
        found = true;
        break;
    case TEMPERA_JSON_NULL:
        found = false;
        break;
    case TEMPERA_JSON_NUMBER:
        out_error->message = tempera_strdup(error->text);
        found = true;
        break;
    case TEMPERA_JSON_BOOL:
        out_error->message = tempera_strdup(error->bool_value ? "true" : "false");
        found = true;
        break;
    case TEMPERA_JSON_ARRAY:
        out_error->message = tempera_strdup("");
        found = true;
        break;
    }
    tempera_json_free(root);
    if (found && out_error->message == NULL) {
        /* The only failure mode left is an allocation failure. */
        return false;
    }
    return found;
}
